pandas/core/groupby/groupby.py

```diff
@@ -1707,7 +1708,12 @@ def nth(self,
             if not self.as_index:
                 return out
 
-            out.index = self.grouper.result_index[ids[mask]]
+            result_index = self.grouper.result_index
+            out.index = result_index[ids[mask]]
+
+            if not self.observed and isinstance(
+                    result_index, CategoricalIndex):
+                out = out.reindex(result_index)
```

Comment:
Also note that this reindexing only works when the user does not provided a `dropna` argument to `nth`, but that seemed to cause an orthogonal issue I can open later